src/combinatorics/simple_graph/chromatic.lean

```diff
@@ -0,0 +1,190 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller
+-/
+
+import combinatorics.simple_graph.connectivity
+import combinatorics.simple_graph.coloring
+import data.polynomial.eval
+import algebra.big_operators.basic
+import order.upper_lower
+
+/-!
+# Chromatic polynomial
+
+-/
+
+open_locale big_operators
+open_locale classical
+
+namespace simple_graph
+variables {V : Type*} (G : simple_graph V)
+
+noncomputable
+def chromatic_polynomial [fintype V] (R : Type*) [ring R] : polynomial R :=
+∑ H in set.to_finset {H | H ≤ G},
+  (-1) ^ fintype.card H.edge_set *
+  polynomial.X ^ fintype.card H.connected_component
+
+-- Direct calculation.
+-- note: should be true for `[ring R]` as well since calculation takes place in center.
```

Comment:
Why not just define it over `ℤ` and use [polynomial.map](https://leanprover-community.github.io/mathlib_docs/data/polynomial/eval.html#polynomial.map) if you want other rings? If you want maximal generality you could do
```
def chromatic_function [fintype V] {R : Type*} [ring R] (r : R) : R :=
∑ H in set.to_finset {H | H ≤ G},
  (-1) ^ fintype.card H.edge_set *
  r ^ fintype.card H.connected_component
```
as well.

Reply:
I could also ask "why not just define it over `R`?" What I was thinking was that when you expose the ring you can write lemmas to say how it interacts with `polynomial.map`, and you can also give `ℤ` as a default argument for `R` (this is a feature that should work better in Lean 4, and I didn't spend time testing it out here).

Regarding `chromatic_function`, I'm not sure along which axis you're considering this to be more general, but sort of the point of this is to give a *polynomial* that happens to interpolate the values of the number of colorings for each cardinality. You'd need to write a lemma that says evaluating it at `polynomial.X` then evaluating it at `r` you get the same function, which seems like a roundabout way to give a polynomial. (Nomenclature-wise, note that there is also the "chromatic (symmetric) function" in the literature.)

Reply:
> Regarding chromatic_function, I'm not sure along which axis you're considering this to be more general,

`chromatic_polynomial R` is defeq to `chromatic_function (X : R[X])` right? I just want to note that your definition is purely in terms of ring operations so it generalize to rings, not just polynomial rings. [polynomial.monomial](https://leanprover-community.github.io/mathlib_docs/data/polynomial/basic.html#polynomial.monomial) would be a spelling that doesn't work over general rings, but the `C r * X ^ n` spelling seems much more common in mathlib.

> I could also ask "why not just define it over R?"

It's your choice! I just want to note that you can prove it over ℤ (which is commutative) and then transfer the result to not-necessarily-commutative rings via `polynomial.map` `and [int.cast_ring_hom](https://leanprover-community.github.io/mathlib_docs/data/int/cast/lemmas.html#int.cast_ring_hom).

Reply:
> I just want to note that your definition is purely in terms of ring operations so it generalizes to rings, not restricted to polynomial rings.

What I was getting at with "You'd need to write a lemma" is that writing a definition using `polynomial R` is a proof that the function uses a fixed "program" of ring operations that does not change depending on the input. It's true that your `chromatic_function` is a polynomial at the meta level, but where is the Lean-level proof that it is one? Writing the definition as a polynomial easily handles this.

This is sort of like how one could write a function `a -> b` and then prove that there is an equivalence, or you could just write the equivalence. In this case, rather than writing a function `R -> R`, you go ahead and write the `polynomial R` (which is a special kind of function `R -> R`).

Regarding the calc block being over a somewhat arbitrary ring, you are completely right that it can be proved over `int` and then boosted to arbitrary rings (and thanks for pointing out a particular way to do this). Stylistically, I think it's good to prove things where concrete types are replaced by type variables (in this case `R` for `int`) since then that guarantees the proof makes no use of the structure of terms of `int`. It gives a loose guarantee that the proof only uses facts about commutative rings.